test/extended/single_node/topology.go

```diff
@@ -0,0 +1,148 @@
+	Expect(err).NotTo(HaveOccurred())
+
+	return list.Items
+}
+
+func getTopologies(f *e2e.Framework) (controlPlaneTopology, infraTopology v1.TopologyMode) {
+	oc := exutil.NewCLIWithFramework(f)
+	infra, err := oc.AdminConfigClient().ConfigV1().Infrastructures().Get(context.Background(),
+		"cluster", metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+
+	return infra.Status.ControlPlaneTopology, infra.Status.InfrastructureTopology
+}
+
+func isInfrastructureDeployment(deployment appsv1.Deployment) bool {
+	infrastructureNamespaces := map[string][]string{
+		"openshift-ingress": {
+			"router-default",
+		},
+	}
+
+	namespaceInfraDeployments, ok := infrastructureNamespaces[deployment.Namespace]
+
+	if !ok {
+		return false
+	}
+
+	for _, infraDeploymentName := range namespaceInfraDeployments {
+		if deployment.Name == infraDeploymentName {
+			return true
+		}
+	}
+
+	return false
+}
+
+func validateReplicas(deployment appsv1.Deployment,
+	controlPlaneTopology, infraTopology v1.TopologyMode, failureAllowed bool) {
+	if isInfrastructureDeployment(deployment) {
+		if infraTopology != v1.SingleReplicaTopologyMode {
+			return
+		}
+	} else if controlPlaneTopology != v1.SingleReplicaTopologyMode {
+		return
+	}
+
+	Expect(deployment.Spec.Replicas).ToNot(BeNil())
+
+	replicas := int(*deployment.Spec.Replicas)
+
+	if !failureAllowed {
+		Expect(replicas).To(Equal(1),
+			"%s in %s namespace has wrong number of replicas", deployment.Name, deployment.Namespace)
+	} else {
+		if replicas == 1 {
+			t := GinkgoT()
+			t.Logf("Deployment %s in namespace %s has one replica, consider taking it off the topology allow-list",
+				deployment.Name, deployment.Namespace)
+		}
+	}
+}
+
+func isAllowedToFail(deployment appsv1.Deployment) bool {
+	// allowedToFail is a list of deployments that currently have 2 replicas even in single-replica
+	// topology deployments, because their operator has yet to be made aware of the new API.
+	// We will slowly remove deployments from this list once their operators have been made
+	// aware until this list is empty and this function will be removed.
+	allowedToFail := map[string][]string{
+		"openshift-authentication": {
+			"oauth-openshift",
+		},
+		"openshift-console": {
+			"console",
+			"downloads",
+		},
+		"openshift-monitoring": {
+			"prometheus-adapter",
+			"thanos-querier",
+		},
+		"openshift-operator-lifecycle-manager": {
+			"packageserver",
+		},
+	}
+
+	namespaceAllowedToFailDeployments, ok := allowedToFail[deployment.Namespace]
+
+	if !ok {
+		return false
+	}
+
+	for _, allowedToFailDeploymentName := range namespaceAllowedToFailDeployments {
+		if deployment.Name == allowedToFailDeploymentName {
+			return true
+		}
+	}
+
+	return false
+}
+
+var _ = Describe("[sig-arch] Cluster topology single node tests", func() {
+	f := e2e.NewDefaultFramework("single-node")
+
+	It("Verify that all OpenShift namespaces deployments take control plane topology into account", func() {
```

Comment:
Verify that OpenShift components deploy one replica in single node mode

Reply:
what is "all" ?

Reply:
> what is "all" ?

All deployments defined inside `openshift-` namesapces

Reply:
Re-phrased to "Verify that OpenShift components deploy one replica in SingleReplica topology mode" single it's not called single-node in the API topology enum